docs/tools/build_rules.md

```diff
@@ -0,0 +1,118 @@
+
+ * `TARGET`, constructed from the configuration value `target.extra_labels` and the name of the targets.
+ * `FEATURE`, constructed from the configuration value `target.features`.
+ * `COMPONENT`, controlled from the configuration value `target.components`.
+ * `TOOLCHAIN`, controlled completely by the toolchain used to build.
+
+
+#### Target Directories
+
+The target labeled directories are used for selecting sources that are specific to an MCU or MCU family. The set of directories included in a given build is all of the names of the targets in the inheritance heirachy and the `target.extra_labels`. The following is an example `targets.json`:
+
+```json
+{
+    "FREESCALE": {
+        "inherits": ["Target"],
+        "public": false
+    },
+    "TEENSY3_1": {
+        "inherits": ["FREESCALE"],
+        "extra_labels": ["K20XX", "K20DX256"]
+    }
+}
+```
+
+The above example will include source files in directories named `TARGET_FREESCALE` and `TARGET_TEENSY3_1`, from the inheritance heirachy and `TARGET_K20XX` and `TARGET_K20DX256` from the `target.extra_labels` value. It will not include source files within `TARGET_NORDIC`, `TARGET_K66F` or `TARGET_NUCLEO_F411` directories as they are not in the list of labels included in the label type `TARGET`
+
+#### Feature Directories
+
+The feature labeled directories are used for software that implements some funtionality that requires some amount of space even when it's not used. The set of directories include by the `FEATURE` label type is entirely controlled by the configuration value `target.features`. The following is a trimmed down version of an example `targets.json`:
+
+```json
+{
+    "NRF52_DK": {
+        "inherits": ["Target"],
+        "features": ["BLE"]
+    }
+}
+```
+
+In the above example, `mbed compile` will include files in directories named `FEATURE_BLE`, and not directories such as `FEATURE_STORAGE` or `FEATURE_CRYTOCELL310`.
+
+#### Component Directories
+
+The component labeled directories are used for software that implements some funtionality. They are within label directory primarily because we don't expect every program to use this software, and including this software in every build would needlessly increase build time.  The set of directories include by the `COMPONENT` label type is entirely controlled by the configuration value `target.components`. The following is a trimmed down version of an example `targets.json`:
+
+```json
+{
+    "NRF52_DK": {
+        "inherits": ["Target"],
+        "components": ["SPIF"]
+    }
+}
+```
+
+In the above example, `mbed compile` will include files in directories named `COMPONENT_SPIF`, and not directories such as `COMPONENT_SD` or `COMPNONENT_FLASHIAP`.
+
+#### Toolchain Directories
+
+The toolchain labeled directories are used for toolchain specific files such as assembly or linker files. The compilers use the following label sets:
+
```

Comment:
I think this is more readable if alphabetically sorted by Toolchain

Reply:
@TacoGrandeTX It helps if you respond to the first/second line in the table (instead of the lines above it) as it makes it possible to see what you're talking about, and it will collapse when I change the order.

Reply:
fixed by 60dfad4